### DISCUSSION TITLE
Upgrade Instructions

### DISCUSSION THREAD

**Discussion by alexander-rieder** [ngoduykhanh/PowerDNS-Admin#1523] [Ideas]
Hello Everyone,

what are the recommended steps for updating an installation without docker from October 2019?

Thanks for your help,
Alex

**Comment by johnwc** [comment]
I would like this as well. Also would be nice to know what version of PDNS it has been tested to work with.

**Comment by falz** [comment]
Still chiming in the lack of upgrade instructions on the wiki, which appears to have been rewritten:

https://github.com/PowerDNS-Admin/PowerDNS-Admin/tree/master/docs/wiki

Unsure why this was removed from the milestone for .4 - because it wasnt completed?

There are several open tickets regarding issues related to things such as quotes in the changelog (https://github.com/PowerDNS-Admin/PowerDNS-Admin/issues/1229)  and a repeat of this ticket asking for upgrade docs (https://github.com/PowerDNS-Admin/PowerDNS-Admin/issues/884)

Curious if anyone has general clues on the best method to upgrade. We plan on installing it into a different host machine using existing SQL data, possibly in a different database as a backup in case of issues.

**Comment by AzorianMatt** [comment]
@falz as for why it hasn't been done yet, I would say it's because the project ran into a maintenance issue last year and is only just now starting to get back on track. I removed it from the latest release milestone because it was not ready yet but I wanted to get the release out on the promised timeline.

It won't be forgotten, but I can't make any promises to delivery dates as of yet.

**Comment by subbink** [comment]
I can have a look and see if I can make some documentation on how to upgrade.

**Comment by falz** [comment]
Thank you, even a summary of suggested steps in this ticket would be helpful

**Comment by falz** [comment]
Just making some noise to see if anyone has started any upgrade docs yet? It's great to see new releases coming out (0.41 just came out) but we're stuck / not comfortable upgrading until we can see at least an outline of steps.

**Comment by AzorianMatt** [comment]
@falz No one is currently working on this that I'm aware of. Unfortunately, I am not making it a priority for myself as I have much higher priorities than this at this time. If another contributor manages to find the time and interest in working on it, I'll be sure to recommend it.

**Comment by falz** [comment]
I realize folks are busy, but it's a chicken and egg thing. It's great that new releases are being made but.. if people don't know how to upgrade to it (properly) fewer will be upgrading to said new releases and less feedback and testing will occur with new version.

I don't currently have a dev environment setup with my older version, if i did i would do some testing and write something down. 

I took internal notes from the 'before times' for upgrade, placing here for now. This is from this link, which now appears to be dead https://github.com/ngoduykhanh/PowerDNS-Admin/wiki/Running-PowerDNS-Admin-on-Ubuntu-or-Debian

per my internal instructions below, i run it as the **powerdnsadmin** local user, and it's installed in **/opt/powerdns-admin**. As one can see below this was off of the git version, which we probably should not be using - so somewhere between 0.2.3 and 0.2.4.

```
git log -1
commit 32983635c6bc51d2ca74496aa9cb0cf834fe2563
Author: jyoung15 <jyoung15@users.noreply.github.com>
Date:   Fri May 7 17:43:44 2021 -0400
```

as **root** user
```
systemctl stop powerdns-admin
```

as the powerdnsadmin user (that it runs as, ymmv)
```
cd /opt/powerdns-admin
git pull
 
# activate venv
. ./flask/bin/activate
 
# set config variables, upgrade db
export FLASK_CONF=/opt/powerdns-admin/configs/config.py
export FLASK_APP=powerdnsadmin/__init__.py
flask db upgrade
 
# install static assets
yarn install --pure-lockfile
flask assets build
```

them as **root** user, start service
```
systemctl start powerdns-admin
```

**Reply by AzorianMatt** [reply]
@falz I certainly understand and appreciate your perspective on the matter and don't disagree. However, when considering the long-term goals of the project, over-investment into the legacy edition is not on that list since it will be replaced.

This isn't to say that things like this won't get done, but I simply cannot commit to any such specifics. My primary goal for the legacy edition right now is to achieve feature stability, so that it can idle a bit more while the next-gen application is worked on.

**Reply by falz** [reply]
Just to note i'm not explicitly talking about 'legacy' (pre 0.4.x?) but any version (0.4.0 -> 0.4.1?) . cheers, hoping this isn't coming off too harshly :grimacing:

**Reply by AzorianMatt** [reply]
@falz I completely understand. Things like this will be handled much better in the next app, but I don't expect many of the issues to be cleaned up on the current one. It's certainly not in my intention for the legacy edition to see many more releases, especially those which would require special upgrade instructions.

For what it's worth, I'm planning a completely interactive tool in the new app to import data from the legacy edition. That should make the upgrade there rather smooth for many people.

**Reply by falz** [reply]
I think i'm confused on something. what is the 'next app'? Is this project that i'm commenting on 'legacy edition', all versions and there's a new separate fork?

**Reply by AzorianMatt** [reply]
@falz The next generation app will be built on a different framework (Python based) to replace the Legacy edition which is the current application and this project. It will be its own separate project and you can track its progress under this organization in a different repository.

**Reply by falz** [reply]
Does the roadmap for the new project include compatability or the ability to cleanly import accounts/users/passwords from odl to new?

**Reply by AzorianMatt** [reply]
@falz yes. I'm planning on building either a CLI or GUI tool to provide import capabilities from the legacy edition where possible.

**Comment by WhatshallIbreaktoday** [comment]
Hello,  fwiw I can share what I do :   I install each new version in it's own directory,  like /opt/powerdns-admin/Powerdns-admin-0.4.0/ by untar ..  use a symlink to target the "current" release I want to use.

This allows me to tweak each release if needed... and roll-back..  (if the database changes allows it) and to keep each version's config file.

Side note :  The wiki page on the Ubuntu install changed a bit for the 0.4.1 (instead of flask activate.. it uses venv...)
Also there are some new packages added with apt..)
Had to changes the gunicorn binary reference in my service file.. but after all everything is working.

I would highly recommend having a test environment to test / fix things before planning to migrate on Prod environment.

Best regards,

**Comment by falz** [comment]
This is what i do as well, almost exactly. However, my main concern is database migrations. I saw/see many tickets with people having DB issues after upgrades and it's currently 100% unclear if there is even some sort of update script that handles db schema changes during the upgrade process? 

DB is what I really need to figure out as that's the part that makes it hard/impossible to 'go back' easily unless I made a db copy for each upgrade until things stabilize.

**Reply by AzorianMatt** [reply]
@falz We use DB migration scripts between each version, so upgrading should that easy in most cases. As for backups, you really should be doing those regardless. The app's version is `0.4.1` which is very young so I would retain caution.

```
source venv/bin/activate
FLASK_APP=powerdnsadmin flask db migrate
FLASK_APP=powerdnsadmin flask assets build
yarn install --pure-lockfile
./run.py
```

**Reply by falz** [reply]
Glad to hear DB migration scripts exist per norms. Is the issue that they did NOT exist to properly upgrade 0.2.x versions (pre maintainer change) to 0.3.x / 0.4.x versions? Or should folks generally expect things to work if they do an upgrade one version at a time?

**Reply by AzorianMatt** [reply]
@falz I'm not highly familiar with the status of things in the early versions but in general, I think jumping versions sequentially should get the job done as an upgrade path. I would assume some form of migration system likely existed from the beginning.
